### DISCUSSION TITLE
feat(add): pypi release

### DISCUSSION THREAD

**Discussion by QIN2DIM** [QIN2DIM/hcaptcha-challenger#18] [Ideas]
We will release the heterogeneous scaffolding in `v1.0` and upload the project package to pypi platform, then players can use `pip` to install and use `hcaptcha-challenger`.

But before that, we will absorb as many suggestions from players as possible, and we will also re-standardize the model of the scaffolding interface so that it can be called in various suitable occasions.

**Comment by GoldDust69** [comment]
Hello, I just wondered if you had any plans to enable multi threading for this. For my custom use case, it seems that the "Image Downloader" makes this incompatible due to the asyncio loop structure. Unfortunately, I do not have the skills to modify this myself and was left wondering if this could be a useful feature for other people too.

**Comment by QIN2DIM** [comment]
> For my custom use case, it seems that the "Image Downloader" makes this incompatible due to the asyncio loop structure. 

？👀

**Comment by GoldDust69** [comment]
> > For my custom use case, it seems that the "Image Downloader" makes this incompatible due to the asyncio loop structure.
> 
> ？👀

For example, when I try the captcha challenge in a function, such as the following:
```
def run_threaded(job_func):
    job_thread = threading.Thread(target=job_func)
    job_thread.start()
    
run_threaded(job_func)
```
I receive the following error: "RuntimeError: There is no current event loop in thread 'Thread-2 (job_func)".

I tried the following for `ImageDownloader()` instead, but it was then not downloading the images:
```
loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)
loop.run_until_complete(downloader.subvert(workers="fast"))
```
Apologize for my lack of experience but hopefully you understand :)

**Comment by QIN2DIM** [comment]
```python
def run_threaded(job_func):
    job_thread = threading.Thread(target=job_func)
    job_thread.start()
    job_thread.join()

run_threaded(job_func)
```

**Comment by QIN2DIM** [comment]
```python

if "win" in sys.platform:
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
    asyncio.run(ImageDownloader(docker=docker_).subvert(workers="fast"))
else:
    loop = asyncio.get_event_loop()
    loop.run_until_complete(
        ImageDownloader(docker=docker_).subvert(workers="fast")
    )
```

**Comment by GoldDust69** [comment]
> ```python
> def run_threaded(job_func):
>     job_thread = threading.Thread(target=job_func)
>     job_thread.start()
>     job_thread.join()
> 
> run_threaded(job_func)
> ```



> ```python
> if "win" in sys.platform:
>     asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
>     asyncio.run(ImageDownloader(docker=docker_).subvert(workers="fast"))
> else:
>     loop = asyncio.get_event_loop()
>     loop.run_until_complete(
>         ImageDownloader(docker=docker_).subvert(workers="fast")
>     )
> ```

Thank you so much for taking the time to look at this for me. Your examples work well and I am able to run the function in a thread now. Unfortunately, my goal doesn't seem possible. I was hoping to be able to run multiple threads simultaneously which is why I never used ```job_thread.join()``` in my previous code.

**Comment by QIN2DIM** [comment]
我认为有多种方案可以实现你所谓的“异步协同”，此外，使用“协程异步”替代“线程异步”能在这类场景中争取到更多的资源利用率。但为了避免所谓的“滥用”我觉得我不能至少不应该在这种场合将“高性能”的并发方案公之于众。如果你想在你自己的项目中使用 hcaptcha-challenger 你应该知道怎么修改代码。

**Comment by GoldDust69** [comment]
> 我认为有多种方案可以实现你所谓的“异步协同”，此外，使用“协程异步”替代“线程异步”能在这类场景中争取到更多的资源利用率。但为了避免所谓的“滥用”我觉得我不能至少不应该在这种场合将“高性能”的并发方案公之于众。如果你想在你自己的项目中使用 hcaptcha-challenger 你应该知道怎么修改代码。

Thank you for your response. I understand what you mean, unfortunately since this is out of my scope I think I'll have to wait until I have further knowledge of Python. None the less, I appreciate you taking the time to reply.

**Comment by QIN2DIM** [comment]
@GoldDust69 纸上得来终觉浅 绝知此事要躬行

**Comment by mewzax** [comment]
> We will release the heterogeneous scaffolding in `v1.0` and upload the project package to pypi platform, then players can use `pip` to install and use `hcaptcha-challenger`.
> 
> But before that, we will absorb as many suggestions from players as possible, and we will also re-standardize the model of the scaffolding interface so that it can be called in various suitable occasions.

ayyyyyyyyyyyyyy thank u so much
